src/main/java/io/cryostat/discovery/DiscoveryStorage.java

```diff
@@ -63,173 +56,142 @@
-        this.persistencePath = persistencePath;
+        this.deployer = deployer;
+        this.builtin = builtin;
+        this.dao = dao;
         this.gson = gson;
         this.http = http;
         this.logger = logger;
     }
 
     @Override
     public void start(Promise<Void> future) throws Exception {
-        for (String s : fs.listDirectoryChildren(persistencePath)) {
-            Path p = persistencePath.resolve(s);
-            try (BufferedReader br = fs.readFile(p)) {
-                PluginInfo pluginInfo = gson.fromJson(br, PluginInfo.class);
-                if (Objects.equals(pluginInfo.getCallback(), NO_CALLBACK)) {
-                    continue;
-                }
-                map.put(UUID.fromString(s), pluginInfo);
-            } catch (IOException ioe) {
-                logger.error(ioe);
-            }
-        }
-
-        map.entrySet()
-                .forEach(
-                        plugin -> {
-                            UUID key = plugin.getKey();
-                            URI uri = plugin.getValue().getCallback();
-                            http.postAbs(uri.toString())
-                                    .timeout(1_000)
-                                    .followRedirects(true)
-                                    .send()
-                                    .onSuccess(
-                                            res -> {
-                                                if (!HttpStatusCodeIdentifier.isSuccessCode(
-                                                        res.statusCode())) {
-                                                    removePlugin(key, uri);
-                                                    return;
-                                                }
-                                                logger.info("Found discovery service {}", uri);
-                                            })
-                                    .onFailure(
-                                            t -> {
-                                                t.printStackTrace();
-                                                removePlugin(key, uri);
-                                            });
-                        });
-        future.complete();
-    }
-
-    private void removePlugin(UUID uuid, URI uri) {
-        map.remove(uuid);
-        try {
-            fs.deleteIfExists(persistencePath.resolve(uuid.toString()));
-        } catch (IOException ioe) {
-            logger.error(ioe);
-        }
-        logger.info("Stale discovery service {} removed", uri);
+        List<Future> futures =
+                dao.getAll().stream()
+                        .map(
+                                plugin -> {
+                                    UUID key = plugin.getId();
+                                    URI uri = plugin.getCallback();
+                                    if (uri == NO_CALLBACK) {
+                                        removePlugin(key, key);
```

Comment:
I think this should be `removePlugin(key, uri)`.

Reply:
In this case the `uri` is `null` (`NO_CALLBACK`). The API handler for registering a plugin does not allow null URIs, so the only way this can happen is for registrations performed in-process by the `BuiltInDiscovery`. It's meaningless for Cryostat to perform a callback startup check to itself, so this null special case is written in here.